### ISSUE TITLE
Write prototype Node backend

### ISSUE THREAD

**Issue by dandelionmane** [sourcecred/sourcecred#24]
## essentials
- should be able to read git data, probably via NodeGit 
  * (@wchargin I know you have some thoughts on this)
- should use the same Flow type signatures as the frontend
- should serve our frontend and supply the data

## features
- we keep a persisted copy of the backend's output in source control (a la src/data.json today), and use that to serve the development frontend
  - in the future case, this will be a db
  - we will update the persisted copy occasionally, maybe once a week
- we should also have an end-to-end script that freshly loads the data from the backend, and serves it on the frontend
  - we can use it for E2E CI testing, assuming it's fast enough
- also, it would be good to turn on backend performance testing early

**Comment by wchargin** [comment]
> * should be able to read git data, probably via NodeGit

I’ve been doing a bunch of hacking with `nodegit` and `simple-git`. Here are my thoughts.

Of the two, `simple-git` (known on GitHub as `steveukx/git-js`) is certainly simpler. It provides a JS veneer over standard Git APIs. But I think that it may be too simple. Most of its commands seem to be wrappers around the actual `git(1)` CLI, whose output is parsed by `simple-git`. But this parsing is flaky—e.g., `repo.log(['--stat'])` silently gives completely wrong output, and `repo.log(['--follow', filename])` has the same issue that `gitpython` does with renames (creating “filenames” like `"file => file2"`).

We might have to consider coming back to `simple-git`, but it’s not my top choice. Combine the fact that it will spawn a ton of subprocesses (not great) with the fact that it doesn’t even do a great job of parsing the output, and I don’t see a compelling reason to use it.

Now, `nodegit` takes a different approach. It exposes direct bindings to `libgit2`, and in fact is the official set of JS bindings for this standard library (as seen on [the `libgit2` homepage][libgit2]). This suggests to me that it should theoretically have better support. In my opinion, the library has two major downsides. First, while the documentation serves as a useful skeleton for the type hierarchies and structures, its explanatory value is next to nothing. For instance, [here’s the page for `Revwalk`][nodegit-revwalk]. I can gather that it’s supposed to walk over revisions, and I can see what _types_ of arguments I’m supposed to pass to the `push*` family of functions—but I have no idea what they actually do. (Why does a file walker need a stack?) The `hide` options are even more confusing.

That said, the failures of the documentation are partially mitigated by the fact that `nodegit`’s architecture very closely mirrors that of `libgit2` itself, and `libgit2`’s documentation is far better. The [`libgit2` documentation for `git_revwalk_push`][libgit2-revwalk-push] answers all my questions in plain English. (Believe it or not, it was easier for me to write a working program using `libgit2` in C than the same program using `nodegit` in JS, given that I did the latter first.)

The other problem that I have with `nodegit` is that… well, I’ve just had an unusual of trouble getting simple examples working. Things like, their docs say that a callback is called with `(null, null)` when iteration finishes, but it’s actually called with `(someErrorObject, undefined)`. I even managed to get a ten-line program, basically copied from their examples, to throw a `SIGABRT` (134) and hard-crash all of node ([issue details])—this worries me a bit.

In summary: I really want to like `nodegit`, but I’m not sure that it’s the right choice, either. Maybe parsing the output of `git(1)` CLI commands really is simpler. I’m open to opinions.

[libgit2]: https://libgit2.github.com/
[nodegit-revwalk]: http://www.nodegit.org/api/revwalk/
[libgit2-revwalk-push]: https://libgit2.github.com/libgit2/#HEAD/group/revwalk/git_revwalk_push
[issue details]: https://github.com/nodegit/nodegit/issues/1442

**Comment by wchargin** [comment]
Note on subprocessing: the forking itself is not _that_ bad. I can batch `vfork`/`exec` with throughput 9200 Hz on my laptop.  With ~one subprocess per repo commit, that makes ~3s overhead for tensorflow/tensorflow (29K commits) and ~5s overhead for git/git (50K commits).

<details>

```c
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

int main() {
    /* Runtime: 1m53.459s = 113.459s = 0.1082 ms/op */
    /* When unbatched: 0.2122ms/op */
    for (size_t i = 0; i < 1024; i++) {
        pid_t pids[1024];
        for (size_t j = 0; j < 1024; j++) {
            pid_t pid = vfork();
            if (!pid) {
                execl("/bin/true", "/bin/true", NULL);
            }
            pids[j] = pid;
        }
        for (size_t j = 0; j < 1024; j++) {
            waitpid(pids[j], NULL, 0);
        }
    }
    return 0;
}
```
</details>

It’s still possible that a subprocess-based solution could be significantly slower, though, because the subprocess can’t share memory, state, or work.

**Comment by dandelionmane** [comment]
NodeGit seems to be basically unmaintained - observe the two month-old PRs with no comments or progress, and very spare activity on the issues. So, given that you're running into so many issues, I think the only feasible way to use this would be to fork it (with an eventual hope of merging all our improvements back in).

SimpleGit seems to have an active owner and maintenance, but as you note it doesn't look like a perfect match for us. If we were to use it, I think it would be another case where we should fork it, fix it up to our needs, but with a better trajectory for merging it back.

So - given that we will be maintaining a Git API, which is best? Fork nodegit, fork simplegit, or roll our own? Rolling our own makes sense to me - we only need a small subset of Git integration; since we only care about reads and not writes, we will focus our attention differently and get the semantics precisely correct for our use case. However, it seems like there are two major approaches: we can wrap the CLI (in which case we may benefit by taking some code from SimpleGit) or we can wrap libgit2 (in which case we may benefit taking some code from NodeGit).

Alternatively, we could take a step back and re-assess whether we want to use Node for the backend, given that when we made the call for Node, we believed we would have high-quality git libraries. But I still think using Node is worth it; being able to share code, types, testing, and infrastructure across the frontend and backend will make maintaining the system a lot easier. Also, I don't see compelling advantages to using any other language (Go or Python) as the backend.

**Comment by dandelionmane** [comment]
The root content of this issue is outdated, since it describes a backend which consists entirely of serving Git data, but now we are pivoting to a pluggable graph-based backend. However, I think the discussion here is still relevant, for the specific question of how to integrate with Git - which makes me think we should either close this (but reference the contents when we develop a git plugin), or cannibalize this thread into something about making a Git plugin.

@wchargin thoughts?

**Comment by dandelionmane** [comment]
I'm going to close this issue as obsolete. We can reference back to it if we need context on building Git integration, although as it looks complicated I think we can punt on it for the near term.
